Main Model/firm-procedures.nls

```diff
@@ -0,0 +1,445 @@
+to decide-fire-worker
+;  (ifelse
+;    desired-labor-change < 0 and n-workers > 1 [  ; firms won't fire their last worker
+;      ask one-of my-employment-links [die]
+;      set months-with-all-positions-filled months-with-all-positions-filled + 1  ; if they fired, this was a month with all positions filled
+;      set desired-labor-change 0  ; no longer want to fire
+;    
+;    desired-labor-change = 0 [
+;      set months-with-all-positions-filled months-with-all-positions-filled + 1
+;    ] [  ; if the firm still wants to hire at the end of the month
+;      set months-with-all-positions-filled 0
+;    ]
+;  )
+
+  if close-position? [  ; firms won't fire their last worker if firm-exit is disabled
+    if n-workers > 0 [
+      ask one-of my-employment-links [die]
+    ; set months-with-all-positions-filled months-with-all-positions-filled + 1  ; if they fired, this was a month with all positions filled
+    ; set desired-labor-change 0  ; no longer want to fire ( I used this before using close-position?
+    ]
+  ]
+  set close-position? false
+
+  ;set-firm-size
+end
+
+;firm procedure-determines whether exit is an appropriate response for the firm this period
+to remove-firm
+  if replace-exited-firm? [
+    let f-type firm-type
+    add-firm f-type
+  ] 
+  ask consumer-link-neighbors [
+    create-consumer-link-with one-of other firms with [consumer-good-firm?][
+      init-consumer-link
+    ]
+  ]
+  
+  die
+end
+
+
+;firm procedure, adds in another firm of a given firm type
+to add-firm [new-firm-type]
+  foreach FIRM-STRUCTURE[f ->
+    let group table:get f "Firm type"
+    if (group = new-firm-type) [
+      hatch 1[
+        set price mean [price] of firms 
+        set wage-rate mean [wage-rate] of firms
+        set liquidity 50 ; model begins at the beginning of a month, so start with the buffer amount
+        raise-funds liquidity
+        set last-open-position -2 * months-to-lower-wage  ; arbitrarily set > γ so firms don't try to immediately lower wage
+        set filled-position? false  ; just needs to be set to a boolean
+        set open-position?  true  ; entering firm would like to hire a new worker
+        set close-position? false  ; not firing immediately by default
+        if input-data != "None" [
+          set demand mean [demand] of firms / 2 ;start smaller
+          set inventory mean [inventory] of firms ; I tried (.5 * demand) to start with inventory such that firms won't want to hire or fire on the first tick, but didn't seem to work
+          set-daily-input-demands
+        ]
+      ]
+    ]
+  ]
+end
+
+;firm procedure, firms search for the funding from households
+to raise-funds [startup-amount]
+  let to-be-raised startup-amount
+  let new-firm self
+  while [to-be-raised > 0] [
+    ask one-of households [   ;guarantees that we do not take the same investor twice
+      let liquidity-available liquidity * 0.5
+      ifelse liquidity-available > to-be-raised [
```

Comment:
there is a bunch of repeated logic in this ifelse. I would just set a local variable in the ifelse for the amount the household is going to invest and then do everything else outside the ifelse (create the equity link, set to-be-rased, set liquidity)